zerver/actions/message_delete.py

```diff
@@ -146,3 +147,33 @@ def do_delete_messages_by_sender(user: UserProfile) -> None:
     )
     if message_ids:
         move_messages_to_archive(message_ids, chunk_size=retention.STREAM_MESSAGE_BATCH_SIZE)
+
+
+def delete_deactivated_user_messages(
+    realm: Realm, user_profile: UserProfile, message_delete_action: list[int]
+) -> None:
+    if Message.NO_DELETE_ACTION in message_delete_action:
+        return
+    user_stream_messages = list(
```

Comment:
Maybe we can avoid this query if only direct messages are to be deleted. Or just fetch all the messages once together and then classify it based on what messages user wants to delete.

Reply:
`do_delete_messages` calls `is_stream_messages` now which uses `self.recipient.type` so it would be good to prefetch `recipient` field using `select_related` in `Message` query.